### PR TITLE
Feature/sc 52

### PR DESCRIPTION
**Pull request by DanCh11**

### CODE REVIEW

**Review by costezki — CHANGES_REQUESTED** [review]

File: ml_experiments/services/pwdb_knn_experiments.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+LAW2VEC_SVM_CATEGORY = 'law2vec/SVM/svm_cateogry.pkl'
+LAW2VEC_SVM_SUBCATEGORY = 'law2vec/SVM/svm_subcateogry.pkl'
+LAW2VEC_SVM_TOM = 'law2vec/SVM/svm_type_of_measure.pkl'
+LAW2VEC_SVM_TG_L1 = 'law2vec/SVM/svm_target_groups_l1.pkl'
+
+
+class KNNPWDBExperiment(BaseExperiment):
+
+    def data_extraction(self, *args, **kwargs):
```
Comment (costezki) [inline ml_experiments/services/pwdb_knn_experiments.py:52]: no, do not overwrite these methods, their implementation shall be inherited from PWDBBaseExperiment.

File: ml_experiments/services/pwdb_knn_experiments.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+        pass
+
+    def data_validation(self, *args, **kwargs):
```
Comment (costezki) [inline ml_experiments/services/pwdb_knn_experiments.py:55]: do not implement

File: ml_experiments/services/pwdb_knn_experiments.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        pass
+
+    def data_preparation(self, *args, **kwargs):
```
Comment (costezki) [inline ml_experiments/services/pwdb_knn_experiments.py:58]: idem.

File: ml_experiments/services/pwdb_knn_experiments.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+        knn_w2v = Pipeline([
+            ("word2vec vectorizer", MeanEmbeddingVectorizer(w2v_dict)),
+            ("KNN", KNC())])
+
+        knn_l2v = Pipeline([
+            ("word2vec vectorizer", MeanEmbeddingVectorizer(l2v_dict)),
+            ("KNN", KNC())])
+
+        fit_w2v_category = knn_w2v.fit(train_test_dataset['X_train'], train_test_dataset['y_train']['Category'])
```
Comment (costezki) [inline ml_experiments/services/pwdb_knn_experiments.py:83]: revise the names of all the variables.

File: ml_experiments/services/pwdb_knn_experiments.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+        pickle_w2v_subcategory = pickle.dumps(fit_w2v_subcategory)
+        pickle_w2v_tom = pickle.dumps(fit_w2v_tom)
+        pickle_w2v_target_groups_l1 = pickle.dumps(fit_w2v_target_groups_l1)
+
+        pickle_l2v_category = pickle.dumps(fit_l2v_category)
+        pickle_l2v_subcategory = pickle.dumps(fit_l2v_subcategory)
+        pickle_l2v_tom = pickle.dumps(fit_l2v_tom)
+        pickle_l2v_target_groups_l1 = pickle.dumps(fit_l2v_target_groups_l1)
+
+        minio_ml_experiments.put_object("word2vec/KNN/knn_cateogry.pkl", pickle_w2v_category)
```
Comment (costezki) [inline ml_experiments/services/pwdb_knn_experiments.py:105]: this shall be done with MLFlow. Start using it

File: ml_experiments/services/pwdb_knn_experiments.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        minio_ml_experiments.put_object("word2vec/KNN/knn_subcateogry.pkl", pickle_w2v_subcategory)
+        minio_ml_experiments.put_object("word2vec/KNN/knn_type_of_measure.pkl", pickle_w2v_tom)
+        minio_ml_experiments.put_object("word2vec/KNN/knn_target_groups_l1.pkl", pickle_w2v_target_groups_l1)
+
+        minio_ml_experiments.put_object("law2vec/KNN/knn_cateogry.pkl", pickle_l2v_category)
+        minio_ml_experiments.put_object("law2vec/KNN/knn_subcateogry.pkl", pickle_l2v_subcategory)
+        minio_ml_experiments.put_object("law2vec/KNN/knn_type_of_measure.pkl", pickle_l2v_tom)
+        minio_ml_experiments.put_object("law2vec/KNN/knn_target_groups_l1.pkl", pickle_l2v_target_groups_l1)
+
+    def model_evaluation(self, *args, **kwargs):
```
Comment (costezki) [inline ml_experiments/services/pwdb_knn_experiments.py:115]: what about it?

File: ml_experiments/services/pwdb_knn_experiments.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+        pass
+
+    def model_validation(self, *args, **kwargs):
```
Comment (costezki) [inline ml_experiments/services/pwdb_knn_experiments.py:118]: this can pass for now.

File: ml_experiments/services/pwdb_knn_experiments.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        pass
+
+
+knn = KNNPWDBExperiment()
```
Comment (costezki) [inline ml_experiments/services/pwdb_knn_experiments.py:122]: what is this? 
delete these two lines/.

File: ml_experiments/services/pwdb_svm_experiment.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        minio_ml_experiments.put_object("law2vec/SVM/svm_cateogry.pkl", pickle_l2v_category)
+        minio_ml_experiments.put_object("law2vec/SVM/svm_subcateogry.pkl", pickle_l2v_subcategory)
+        minio_ml_experiments.put_object("law2vec/SVM/svm_type_of_measure.pkl", pickle_l2v_tom)
+        minio_ml_experiments.put_object("law2vec/SVM/svm_target_groups_l1.pkl", pickle_l2v_target_groups_l1)
+
+    def model_evaluation(self, *args, **kwargs):
+        minio = MinioAdapter(MINIO_URL, MINIO_ACCESS_KEY, MINIO_SECRET_KEY, ML_EXPERIMENTS_BUCKET_NAME)
+        train_test_dataset = pickle.loads(minio.get_object(PWDB_TRAIN_TEST))
+
+        w2v_category = pickle.loads(minio.get_object(WORD2VEC_SVM_CATEGORY))
```
Comment (costezki) [inline ml_experiments/services/pwdb_svm_experiment.py:127]: let's implement a function that takes a model from minio, evaluates it and stores the results in the mlflow.

Comment (costezki) [reply ml_experiments/services/pwdb_svm_experiment.py:127]: Programming challenge: shorten `model_evaluation` method to 5-10 lines max.

File: requirements.txt [hunk 10]
```diff
@@ -16,5 +16,7 @@ mlflow
 nltk
 spacy
 jq
+minio
+apache-airflow
```
Comment (costezki) [inline requirements.txt:20]: mlflow too?

**Review by costezki — COMMENTED** [review]
well done! This is a good start. Let's make some improvements.

File: ml_experiments/services/pwdb_base_experiment.py [hunk 11]
```diff
@@ -2,32 +2,158 @@
...
+        minio.put_object("pwdb_prepared.pkl", convert_to_pickle)
+        # Split data for train, test and word model training
+        independent_data = multi_label_target_groups['Concatenated Data (clean)']
+        label_data = multi_label_target_groups.drop(['Concatenated Data (clean)', 'Target groups'], axis=1)
+        pwdb_word2vec = Word2Vec(independent_data, window=5, min_count=10, size=300)
+        w2v_saving = pickle.dumps(pwdb_word2vec)
+        minio.put_object("word2vec.gensim.model", w2v_saving)
+        x_train, x_test, y_train, y_test = train_test_split(independent_data, label_data,
+                                                            random_state=42, test_size=0.3, shuffle=True)
+        train_test_dict = {"X_train": x_train, "X_test": x_test, "y_train": y_train, "y_test": y_test}
```
[the remark this answers is not in the document]
Comment (DanCh11) [reply ml_experiments/services/pwdb_base_experiment.py:143]: I think here is good, because we don't need to repeat the same action for train-test splitting over again for every algorithm

File: ml_experiments/services/pwdb_base_experiment.py [hunk 12]
```diff
@@ -2,32 +2,158 @@
...
+        convert_to_pickle = pickle.dumps(train_test_dict)
```
[the remark this answers is not in the document]
Comment (DanCh11) [reply ml_experiments/services/pwdb_base_experiment.py:144]: but how we can extract the data from this file and use it in algorithm training?

File: ml_experiments/services/pwdb_knn_experiments.py (hunk 6 above)
Comment (DanCh11) [reply ml_experiments/services/pwdb_knn_experiments.py:115]: i have some questions about model_evaluation part that's why i didn't complete this stet yet

File: requirements.txt (hunk 10 above)
Comment (DanCh11) [reply requirements.txt:20]: it's already in this file

File: ml_experiments/services/pwdb_base_experiment.py [hunk 13]
```diff
@@ -2,32 +2,197 @@
...
+    'Cross-border commuters', 'Disabled workers', 'Employees in standard employment',
+    'Female workers', 'Migrants in employment', 'Older people in employment (aged 55+)',
+    'Other groups of workers', 'Parents in employment', 'Particular professions',
+    'Platform workers', 'Posted workers', 'Refugees in employment', 'Seasonal workers',
+    'Self-employed', 'Single parents in employment', 'The COVID-19 risk group at the workplace',
+    'Undeclared workers', 'Unemployed', 'Workers in care facilities',
+    'Workers in essential services', 'Workers in non-standard forms of employment',
+    'Youth (18-25) in employment']
+
+PWDB_REFACTORING_RULES = '''{
```
Comment (costezki) [inline ml_experiments/services/pwdb_base_experiment.py:53]: like so is much better

File: ml_experiments/services/sc_wrangling/value_replacement.py [hunk 14]
```diff
@@ -0,0 +1,29 @@
+
+from sklearn.preprocessing import LabelEncoder
+
+
+class MultiColumnLabelEncoder:
```
Comment (costezki) [inline ml_experiments/services/sc_wrangling/value_replacement.py:5]: do we still need it? Didn't you discive that this is already implemented by sklearn?

File: tests/minio_connection.py [hunk 15]
```diff
@@ -0,0 +1,32 @@
+
+from pickle import dumps, loads
+
+import pandas as pd
+from pandas import DataFrame
+
+from ml_experiments.adapters.minio_adapter import MinioAdapter
+
+
+class MinioTestFileCreation(MinioAdapter):
```
Comment (costezki) [inline tests/minio_connection.py:10]: 1. rename the class, following conventions from Clean Code CH2
2. this seems to be a Fake Minio Adapter, if so, call it accordingly

File: tests/unit/test_prepare_pwdb_data.py [hunk 16]
```diff
@@ -0,0 +1,36 @@
+
+import requests
+
+from pandas import DataFrame
+
+from tests.minio_connection import MinioTestFileCreation
+from ml_experiments.services.pwdb_base_experiment import PWDBBaseExperiment
+
+MINIO_ACCESS_KEY = '2zVld17bTfKk8iu0Eh9H74MywAeDV3WQ'
```
Comment (costezki) [inline tests/unit/test_prepare_pwdb_data.py:9]: these secrets shall be taken from .env using a config file (implementing our practice)

File: tests/unit/test_prepare_pwdb_data.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+        'Target groups': ['tg_l_one|tg_l_two', 'tg_l_one_first|tg_l_two_second'],
+        'Descriptive Data': [['titleone', 'bione', 'cmone'], ['titletwo', 'bitwo', 'cmtwo']]}
+
+    expected_df = DataFrame(data=expected_data)
+
+    pwdb_experiment = PWDBBaseExperiment(minio, requests)
+
+    test_data_preparation = pwdb_experiment.prepare_pwdb_data(test_dataset)
+
+    assert len(test_data_preparation) == len(expected_df)
```
Comment (costezki) [inline tests/unit/test_prepare_pwdb_data.py:36]: you can also write a few assertions about the content

**Review by costezki — COMMENTED** [review]
well done

File: ml_experiments/services/sc_wrangling/value_replacement.py (hunk 14 above)
Comment (DanCh11) [reply ml_experiments/services/sc_wrangling/value_replacement.py:5]: LabelEncoder only takes a 1-d array as an argument.  It is quite easy to roll own label encoder that operates on multiple columns of our choosing, and returns a transformed dataframe

**Review by costezki — APPROVED** [review]
well done, keep on digging!!!
